{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2737029b",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Goal: An Excel sheet with two lists. Would it be possible to create a volcano plot for both the oxytocin pathway and the calcium signaling pathway for each option? Also, could you create a heatmap for both pathways using these two lists?\n",
    "Author: Carsten Knutsen\n",
    "Date: March 01 2023\n",
    "conda_env: bulk_rnaseq\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "399a979b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "146f3176",
   "metadata": {},
   "outputs": [],
   "source": [
    "DEG_FILE = '/home/carsten/alvira_bioinformatics/myometrium/data/2302_bulk_sequencing/03.Result_X202SC22123874-Z01-F001_Homo_sapiens/Result_X202SC22123874-Z01-F001/4.Differential/1.deglist/REDvsGD/REDvsGD_deg.xlsx'\n",
    "OUTPUT = '/home/carsten/alvira_bioinformatics/myometrium/data/230301_daiana_grant/output'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca5233bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_dict = {'Oxytocin genes':[\n",
    "    'CAMKK1',\n",
    "'GUCY1B2',\n",
    "'NPR1',\n",
    "'PGF',\n",
    "'PGF',\n",
    "'PTGDS',\n",
    "'PTGDS',\n",
    "'PTGES3P2',\n",
    "'PTGES3P2',\n",
    "'ACTA1',\n",
    "'ACTC1',\n",
    "'EGFL6',\n",
    "'GNB1L',\n",
    "'GNB4',\n",
    "'GNG4',\n",
    "'GNPNAT1',\n",
    "'GNPTG',\n",
    "'HBEGF',\n",
    "'KCNH7',\n",
    "'KCNIP4',\n",
    "'MAPK4',\n",
    "'MAPK8IP2',\n",
    "'MAPKBP1',\n",
    "# 'MAPL2P8',# What gene is this supposed to be?\n",
    "'MYH13',\n",
    "'MYH3',\n",
    "'NPPA',\n",
    "'NPPB',\n",
    "'PLCB1',\n",
    "'PLCD1',\n",
    "'PRKAA2',\n",
    "'PRKG1',\n",
    "'RGS2',\n",
    "'RGS2',\n",
    "'RGS22',\n",
    "'RGS22',\n",
    "'RGS7BP',\n",
    "'RGS7BP',\n",
    "'RHOB',\n",
    "'RYR2',\n",
    "'TGFB1',\n",
    "'TGFB2',\n",
    "'TRPM5'\n",
    "],\n",
    "             'Calcium genes':[\n",
    "    'ATP10D',\n",
    "'ATP8B1',\n",
    "'ATP8B2',\n",
    "'EGR1',\n",
    "'EGR2',\n",
    "'EGR3',\n",
    "'TNC',\n",
    "'MYH13',\n",
    "'MYH3',\n",
    "'MYO1B',\n",
    "'MYO1E',\n",
    "'ACTA1',\n",
    "'ACTC1',\n",
    "'TNNT1',\n",
    "'BAHCC1',\n",
    "'ATF3',\n",
    "'EGR1',\n",
    "'GNG4',\n",
    "'GUCY1B2',\n",
    "'ITPRIPL2',\n",
    "'NOS3',\n",
    "'PDE4C',\n",
    "'PGF',\n",
    "'PLCB1',\n",
    "'PLCD1',\n",
    "'PRKG1',\n",
    "'PTGDS',\n",
    "'RGS2',\n",
    "'RYR2',\n",
    "'TNNT1',\n",
    "'ITPRIP',\n",
    "]\n",
    "            }\n",
    "for key in gene_dict.keys():\n",
    "    gene_dict[key] = sorted(set(gene_dict[key]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e914d24",
   "metadata": {},
   "outputs": [],
   "source": [
    "deg_df = pd.read_excel(DEG_FILE, index_col = 1)\n",
    "count_df = deg_df[deg_df.columns[1:11]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b712335",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_df = deg_df[['pvalue','padj','log2FoldChange']]\n",
    "volcano_df.dropna(how='any',inplace=True)\n",
    "volcano_df['$-Log_{10}$(FDR)'] = volcano_df['padj'].apply(lambda x: -np.log10(x))\n",
    "volcano_df['$Log_{2}$(FC)'] = volcano_df['log2FoldChange']\n",
    "\n",
    "def conditions(s):\n",
    "    if s['log2FoldChange'] > 0.5 and s['padj'] < 0.1:\n",
    "        return 'Upregulated'\n",
    "    elif s['log2FoldChange'] < -0.5 and s['padj'] < 0.1:\n",
    "        return 'Downregulated'\n",
    "    else:\n",
    "        return 'NS'\n",
    "volcano_df['color'] = volcano_df.apply(conditions, axis=1)\n",
    "for key in gene_dict.keys():\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(3, 4))\n",
    "\n",
    "    sns.scatterplot(data=volcano_df,\n",
    "                    y='$-Log_{10}$(FDR)',\n",
    "                    x='$Log_{2}$(FC)',\n",
    "                    hue='color',\n",
    "                    hue_order=['Upregulated', 'NS', 'Downregulated'],\n",
    "                    palette=['Red', 'grey', 'Blue'],\n",
    "                    s=10,\n",
    "                    linewidth=0,\n",
    "                    ax=ax\n",
    "                    )\n",
    "    for gene in gene_dict[key]:\n",
    "        plt.text( volcano_df['$Log_{2}$(FC)'][gene],volcano_df['$-Log_{10}$(FDR)'][gene], gene, size=8)\n",
    "    ax.get_legend().remove()\n",
    "    ax.set_title(f'{key}')\n",
    "    fig.savefig(f'{OUTPUT}/volcano_{key}_fdr.png', dpi=300, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32d9db59",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ls = ['RYR2','TRPV4','OXTR' ]\n",
    "fig, ax = plt.subplots(1, 1, figsize=(3, 4))\n",
    "\n",
    "sns.scatterplot(data=volcano_df,\n",
    "                y='$-Log_{10}$(FDR)',\n",
    "                x='$Log_{2}$(FC)',\n",
    "                hue='color',\n",
    "                hue_order=['Upregulated', 'NS', 'Downregulated'],\n",
    "                palette=['Red', 'grey', 'Blue'],\n",
    "                s=10,\n",
    "                linewidth=0,\n",
    "                ax=ax\n",
    "                )\n",
    "for gene in gene_ls:\n",
    "    plt.text( volcano_df['$Log_{2}$(FC)'][gene],volcano_df['$-Log_{10}$(FDR)'][gene], gene, size=8)\n",
    "ax.get_legend().remove()\n",
    "ax.set_title(f'')\n",
    "fig.savefig(f'{OUTPUT}/volcano_{key}_custom_fdr.png', dpi=300, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a51d1924",
   "metadata": {},
   "outputs": [],
   "source": [
    "volcano_df.head(50\n",
    "               )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26edfbd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in gene_dict.keys():\n",
    "    sub_df = volcano_df.loc[gene_dict[key]]\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(3, 4))\n",
    "\n",
    "    sns.scatterplot(data=sub_df,\n",
    "                    y='$-Log_{10}$(FDR)',\n",
    "                    x='$Log_{2}$(FC)',\n",
    "                    hue='color',\n",
    "                    hue_order=['Upregulated', 'NS', 'Downregulated'],\n",
    "                    palette=['Red', 'grey', 'Blue'],\n",
    "                    s=10,\n",
    "                    linewidth=0,\n",
    "                    ax=ax\n",
    "                    )\n",
    "    for gene in gene_dict[key]:\n",
    "        plt.text( sub_df['$Log_{2}$(FC)'][gene],sub_df['$-Log_{10}$(FDR)'][gene], gene, size=8)\n",
    "    ax.get_legend().remove()\n",
    "    ax.set_title(f'{key}')\n",
    "    fig.savefig(f'{OUTPUT}/volcano_{key}_fdr_sub.png', dpi=300, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e9d7559",
   "metadata": {},
   "outputs": [],
   "source": [
    "deg_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cb38b74",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in gene_dict.keys():\n",
    "    heatmap_df = count_df.loc[gene_dict[key]]\n",
    "    heatmap_df.columns = ['PC-1','PC-2','PC-3','PC-4','PC-5','GC-1','GC-2','GC-3','GC-4','GC-5']\n",
    "    g = sns.clustermap(data=heatmap_df,\n",
    "                   z_score = 0,\n",
    "                    col_cluster=False,\n",
    "                       square=True,\n",
    "                       yticklabels=True,\n",
    "                   cmap = 'RdBu_r',\n",
    "                       center=0,\n",
    "                       cbar_pos=(0.95,0.3,0.05,0.25),\n",
    "                       figsize=(4, 8)\n",
    "                    )\n",
    "    g.ax_heatmap.set_ylabel(\"\")\n",
    "    g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_ymajorticklabels(), fontsize = 8)\n",
    "    g.ax_cbar.set_ylabel('Z-score')\n",
    "    g.ax_heatmap.set_title(f'{key}')\n",
    "    plt.savefig(f'{OUTPUT}/clustermap_{key}.png', dpi=300, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6f0bd1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in gene_dict.keys():\n",
    "    heatmap_df = count_df.loc[gene_dict[key]]\n",
    "    heatmap_df = np.log10(heatmap_df+1)\n",
    "    heatmap_df.columns = ['PC-1','PC-2','PC-3','PC-4','PC-5','GC-1','GC-2','GC-3','GC-4','GC-5']\n",
    "    g = sns.clustermap(data=heatmap_df,\n",
    "                    col_cluster=False,\n",
    "                       square=True,\n",
    "                       yticklabels=True,\n",
    "                   cmap = 'magma',\n",
    "                       cbar_pos=(0.95,0.3,0.05,0.25),\n",
    "                       figsize=(4, 8)\n",
    "                    )\n",
    "    g.ax_heatmap.set_ylabel(\"\")\n",
    "    g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_ymajorticklabels(), fontsize = 8)\n",
    "    g.ax_cbar.set_ylabel('Log10(MoR)')\n",
    "    g.ax_heatmap.set_title(f'{key}')\n",
    "    plt.savefig(f'{OUTPUT}/clustermap_{key}_log.png', dpi=300, bbox_inches=\"tight\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
