{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45e61059-fcfd-4902-94fe-ae7eec50246d",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Goal:Daina questions on snRNA\n",
    "Author:Carsten Knutsen\n",
    "Date:231114\n",
    "conda_env:uterus_sc\n",
    "Notes: Create figures for Daiana\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a02f3e6-8db8-4501-bba8-8e8714429c2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "# Only needed for processing\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os \n",
    "pd.set_option('display.max_rows', 500)\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "output = '/home/carsten/alvira_bioinformatics/uterus/data/pilot/231114_snRNA_daiana_plan'\n",
    "os.makedirs(output, exist_ok=True)\n",
    "sc.set_figure_params(dpi_save = 300, fontsize=10)\n",
    "sc.settings.figdir = output\n",
    "plt.rcParams[\"font.family\"] = \"Arial\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8b292ed-173d-4671-ba28-0f967e3ce5e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read('/home/carsten/alvira_bioinformatics/uterus/data/single_cell_files/scanpy_files/uterus_processed_celltyped.gz.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc5e5d0c",
   "metadata": {},
   "source": [
    "#### 1) Cell Type Identification:\n",
    "#### Please, could you generate a t-SNE or UMAP plot based on the snRNA-seq data to visualize the clustering of cells and the sub-clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a790bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = 'Cell Subtype', alpha=0.5, save='_Cell Subtype.png')\n",
    "sc.pl.umap(adata, color = 'GroupContract',title='Group', alpha=0.5, save='_Group.png')\n",
    "for group in adata.obs['GroupContract'].cat.categories:\n",
    "    sc.pl.umap(adata, color = 'GroupContract',title='Group',groups=group, alpha=0.5, save=f'_group_{group}.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fa7df15",
   "metadata": {},
   "source": [
    "#### 2) Gene Expression Heatmaps:\n",
    "#### Is it possible to create heatmaps displaying the expression levels of specific gene expressions across different cell clusters to identify potential subpopulations with distinct expression profiles?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e61daa0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''TRPV4: Transient Receptor Potential Vanilloid 4\n",
    "OXTR: Oxytocin Receptor\n",
    "RYR2: Ryanodine Receptor 2\n",
    "ARRb1: Beta-Arrestin 1\n",
    "ARRb2: Beta-Arrestin 2\n",
    "AIP4: Atrophin-Interacting Protein 4 (ITCH - E3 ubiquitin-protein ligase Itchy homolog)\n",
    "IP3: Inositol 1,4,5-Trisphosphate\n",
    "ITPR1: Inositol 1,4,5-Trisphosphate Receptor Type 1\n",
    "ESR1: Estrogen Receptor 1\n",
    "\n",
    "nNOS: Neuronal Nitric Oxide Synthase\n",
    "iNOS: Inducible Nitric Oxide Synthase\n",
    "eNOS: Endothelial Nitric Oxide Synthase\n",
    "\n",
    "TLR1: Toll-like receptors 1\n",
    "TLR2: Toll-like receptors 2\n",
    "TLR3: Toll-like receptors 3\n",
    "NFκB: Nuclear Factor Kappa B\n",
    "TNF-alpha: Tumor Necrosis Factor Alpha\n",
    "\n",
    "VEGFA: Vascular Endothelial Growth Factor A\n",
    "VEGFR: Vascular Endothelial Growth Factor Receptors\n",
    "FGF2: Fibroblast Growth Factor 2 \n",
    "ANGPT1: Angiopoietin 1\n",
    "ANGPT2: Angiopoietin 2\n",
    "NOTCH1: Notch Receptor \n",
    "PDGFB: Platelet-Derived Growth Factor B\n",
    "NRP1: Neuropilin 1'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79e50443",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ['TRPV4',\n",
    "         'OXTR',\n",
    "         'RYR2',\n",
    "         'ARRB1',\n",
    "         'ARRB2',\n",
    "         'ITCH',#'AIP4',\n",
    "         'ITPR3',#'IP3',\n",
    "         'ITPR1',\n",
    "         'ESR1',\n",
    "         'NOS1',\n",
    "         'NOS2',\n",
    "         'NOS3',\n",
    "         'TLR1',\n",
    "         'TLR2',\n",
    "         'TLR3',\n",
    "         'NFKB1',\n",
    "         'NFKB2',\n",
    "         'RELA',\n",
    "         'RELB',\n",
    "         'REL',\n",
    "         'TNF',\n",
    "         'VEGFA',\n",
    "         'VEGFC',\n",
    "         'FLT1',\n",
    "         'FLT4',\n",
    "         'KDR',\n",
    "         'FGF2',\n",
    "         'ANGPT1',\n",
    "         'ANGPT2',\n",
    "         'NOTCH1',\n",
    "         'PDGFB',\n",
    "         'NRP1',#'NPP1'\n",
    "        ]\n",
    "sc.pl.dotplot(adata, genes,groupby='Cell Subtype',standard_scale='var',dendrogram=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2f9e3bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ['TRPV4',\n",
    "         'OXTR',\n",
    "         'RYR2',\n",
    "         'ARRB1',\n",
    "         'ARRB2',\n",
    "         'ITCH',#'AIP4',\n",
    "         'ITPR3',#'IP3',\n",
    "         'ITPR1',\n",
    "         'ESR1',\n",
    "         'NOS1',\n",
    "         'NOS2',\n",
    "         'NOS3',\n",
    "         'TLR1',\n",
    "         'TLR2',\n",
    "         'TLR3',\n",
    "         'NFKB1',\n",
    "         'NFKB2',\n",
    "         'RELA',\n",
    "         'RELB',\n",
    "         'REL',\n",
    "         'TNF',\n",
    "         'VEGFA',\n",
    "         'VEGFC',\n",
    "         'FLT1',\n",
    "         'FLT4',\n",
    "         'KDR',\n",
    "         'FGF2',\n",
    "         'ANGPT1',\n",
    "         'ANGPT2',\n",
    "         'NOTCH1',\n",
    "         'PDGFB',\n",
    "         'NRP1',#'NPP1'\n",
    "        ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46734023",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = sc.get.obs_df(adata, genes+['Cell Subtype'])\n",
    "df2 = df.groupby('Cell Subtype')[df.columns.tolist()[:-1]].mean()\n",
    "cg = sns.clustermap(df2, square=True,standard_scale=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5692233e",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = cg.ax_heatmap.xaxis.get_majorticklabels()\n",
    "genes_sorted = [x.get_text() for x in labels]\n",
    "sc.pl.dotplot(adata, genes_sorted,groupby='Cell Subtype',standard_scale='var',dendrogram=True, save = 'question2_gene_expression_sorted.png')\n",
    "sc.pl.matrixplot(adata, genes_sorted,groupby='Cell Subtype',standard_scale='var',dendrogram=True, save = 'question2_gene_expression_sorted.png')\n",
    "sc.pl.dotplot(adata, genes,groupby='Cell Subtype',standard_scale='var',dendrogram=False, save = 'question2_gene_expression_sorted.png')\n",
    "sc.pl.matrixplot(adata, genes,groupby='Cell Subtype',standard_scale='var',dendrogram=False, save = 'question2_gene_expression_sorted.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d4556c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in genes:\n",
    "    sc.pl.umap(adata, color=gene,alpha=0.3,save=f'_{gene}.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af36d7cd",
   "metadata": {},
   "source": [
    "#### 3) Differential Gene Expression Analysis:\n",
    "#### Please, could you perform a differential gene expression analysis between relevant cell clusters, focusing on genes associated with\n",
    "    • uterine contractility and calcium signaling\n",
    "    • lipid metabolism\n",
    "    • angiogenesis pathway\n",
    "    • NO production\n",
    "#### And generate volcano plots or bar graphs to highlight significantly upregulated or downregulated genes.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75ef6ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Here is a list of some genes per pathway:\n",
    "\n",
    "    a) Uterine Contractility and Calcium signaling\n",
    "Oxytocin Pathway:\n",
    "OXTR: Oxytocin Receptor\n",
    "GNAQ: Guanine Nucleotide-Binding Protein G(q) Subunit Alpha\n",
    "PLCβ: Phospholipase C Beta\n",
    "PKC: Protein Kinase C\n",
    "CaMKII: Calcium/Calmodulin-Dependent Protein Kinase II\n",
    "MAPK: Mitogen-Activated Protein Kinase\n",
    "PI3K: Phosphoinositide 3-Kinase\n",
    "AKT1: AKT Serine/Threonine Kinase 1\n",
    "MYLK: Myosin Light Chain Kinase\n",
    "CALM: Calmodulin\n",
    "ADRA1A: Adrenoceptor Alpha 1A\n",
    "ADRB2: Adrenoceptor Beta 2\n",
    "SRC: SRC Proto-Oncogene, Non-Receptor Tyrosine Kinase\n",
    "EGFR: Epidermal Growth Factor Receptor\n",
    "nNOS: Neuronal Nitric Oxide Synthase\n",
    "iNOS: Inducible Nitric Oxide Synthase\n",
    "eNOS: Endothelial Nitric Oxide Synthase\n",
    "\n",
    "TRPV4 Pathway\n",
    "TRPV4: Transient Receptor Potential Vanilloid 4\n",
    "GNAQ: Guanine Nucleotide-Binding Protein G(q) Subunit Alpha\n",
    "PLCβ: Phospholipase C Beta\n",
    "CaMKII: Calcium/Calmodulin-Dependent Protein Kinase II\n",
    "PKC: Protein Kinase C\n",
    "SRC: SRC Proto-Oncogene, Non-Receptor Tyrosine Kinase\n",
    "AKT1: AKT Serine/Threonine Kinase 1\n",
    "PI3K: Phosphoinositide 3-Kinase\n",
    "MAPK: Mitogen-Activated Protein Kinase\n",
    "EGFR: Epidermal Growth Factor Receptor\n",
    "NFAT: Nuclear Factor of Activated T Cells\n",
    "PTK2: Protein Tyrosine Kinase 2 (also known as FAK, Focal Adhesion Kinase)\n",
    "CAM: Calmodulin\n",
    "RhoA: Ras Homolog Family Member A\n",
    "\n",
    "\n",
    "\n",
    "    b) Lipid metabolism\n",
    "PPARG: Peroxisome Proliferator-Activated Receptor Gamma\n",
    "PPARA: Peroxisome Proliferator-Activated Receptor Alpha\n",
    "PPARD: Peroxisome Proliferator-Activated Receptor Delta\n",
    "RXRA: Retinoid X Receptor Alpha\n",
    "CPT1A: Carnitine Palmitoyltransferase 1\n",
    "FABP4: Fatty Acid Binding Protein 4\n",
    "LPL: Lipoprotein Lipase\n",
    "ACACA: Acetyl-CoA Carboxylase Alpha\n",
    "ACADM: Acyl-CoA Dehydrogenase, C-4 to C-12 Straight Chain\n",
    "SREBF1: Sterol Regulatory Element-Binding Transcription Factor 1\n",
    "DGAT1: Diacylglycerol O-Acyltransferase 1\n",
    "LXR: Liver X Receptor\n",
    "\n",
    "\n",
    "    c) Angiogenesis pathway\n",
    "VEGFA: Vascular Endothelial Growth Factor A\n",
    "VEGFR: Vascular Endothelial Growth Factor Receptors\n",
    "FGF2: Fibroblast Growth Factor 2 \n",
    "ANGPT1: Angiopoietin 1\n",
    "ANGPT2: Angiopoietin 2\n",
    "NOTCH1: Notch Receptor \n",
    "PDGFB: Platelet-Derived Growth Factor B\n",
    "NRP1: Neuropilin 1\n",
    "HIF1A: Hypoxia-Inducible Factor 1 Alpha\n",
    "PECAM1: Platelet and Endothelial Cell Adhesion Molecule 1\n",
    "DLL4: Delta-like ligand 4\n",
    "\n",
    "\n",
    "\n",
    "    d) Nitric oxide production\n",
    "nNOS: Neuronal Nitric Oxide Synthase\n",
    "iNOS: Inducible Nitric Oxide Synthase\n",
    "eNOS: Endothelial Nitric Oxide Synthase\n",
    "GUCY1A3: Guanylate Cyclase 1, Soluble, Alpha 3\n",
    "PRKG1: Protein Kinase, cGMP-Dependent, Type I\n",
    "CaMKII: Calcium/Calmodulin-Dependent Protein Kinase II\n",
    "CA2: Carbonic Anhydrase 2\n",
    "ARG1: Arginase 1\n",
    "HSP90: Heat Shock Protein 90\n",
    "CAT: Catalase\n",
    "SOD: Superoxide Dismutase (there are different isoforms such as SOD1 and SOD2)\n",
    "eNOSIP: Endothelial Nitric Oxide Synthase Interacting Protein\n",
    "DPP4: Dipeptidyl Peptidase 4\n",
    "AKT1: AKT Serine/Threonine Kinase 1\n",
    "RXRA: Retinoid X Receptor Alpha'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d20822bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways = {'Oxytocin Pathway':\n",
    "['OXTR',\n",
    "'GNAQ',\n",
    "'PLCB1',\n",
    " 'PLCB2',\n",
    " 'PLCB3',\n",
    " 'PLCB4',#: Phospholipase C Beta\n",
    "'PKC',#: Protein Kinase C\n",
    "'CaMKII',#: Calcium/Calmodulin-Dependent Protein Kinase II\n",
    "'MAPK',#: Mitogen-Activated Protein Kinase\n",
    "'PI3K',#: Phosphoinositide 3-Kinase\n",
    "'AKT1',#: AKT Serine/Threonine Kinase 1\n",
    "'MYLK',#: Myosin Light Chain Kinase\n",
    "'CALM',#: Calmodulin\n",
    "'ADRA1A',#: Adrenoceptor Alpha 1A\n",
    "'ADRB2',#: Adrenoceptor Beta 2\n",
    "'SRC',#: SRC Proto-Oncogene, Non-Receptor Tyrosine Kinase\n",
    "'EGFR',#: Epidermal Growth Factor Receptor\n",
    "'NOS1',#: Neuronal Nitric Oxide Synthase\n",
    "'NOS2',#: Inducible Nitric Oxide Synthase\n",
    "'NOS3',#: Endothelial Nitric Oxide Synthase}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b34e249f",
   "metadata": {},
   "source": [
    "#### 4) Pathway Analysis:\n",
    "#### Perform pathway analysis to identify enriched biological pathways within specific cell types."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da458802",
   "metadata": {},
   "source": [
    "#### Will run pathway analysis using metascape on cell type markers in each group"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30c6cc20",
   "metadata": {},
   "source": [
    "#### 5) Network Analysis:\n",
    "#### Can we construct gene co-expression networks to reveal potential interactions between genes involved in uterine contractility and calcium signaling, TRPV4 or OXTR modulation, and NO signaling?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23623a9f",
   "metadata": {},
   "source": [
    "#### 6) Cell Trajectory Analysis:\n",
    "#### Do you think that we can create pseudotime plots to illustrate the predicted developmental trajectory of cells within the myometrial tissue?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a468f6f",
   "metadata": {},
   "source": [
    "#### 7) Gene-Set Enrichment Analysis:\n",
    "#### Please, could you perform GSEA to determine whether predefined gene sets related to uterine function, calcium regulation, or NO signaling are enriched in specific cell clusters?\n",
    "\n",
    "* Uterine function during gestation:\n",
    "* ESR1: Estrogen Receptor 1\n",
    "* PGR: Progesterone Receptor\n",
    "* OTX2: Orthodenticle Homeobox 2\n",
    "* PTGES: Prostaglandin E Synthase\n",
    "* ITGB3: Integrin Subunit Beta 3\n",
    "* LHCGR: Luteinizing Hormone/Chorionic Gonadotropin Receptor\n",
    "* IGF1 Insulin-Like Growth Factor 1\n",
    "* VDR Vitamin D Receptor\n",
    "* PLAU Plasminogen Activator, Urokinase\n",
    "* HOXA10: Homeobox A10\n",
    "* IL6: Interleukin 6\n",
    "* COL1A1: Collagen Type I Alpha 1 Chain\n",
    "* CYP19A1: Aromatase\n",
    "* HAND2: Heart and Neural Crest Derivatives Expressed \n",
    "* TFPI: Tissue Factor Pathway Inhibitor\n",
    "* MMP9: Matrix Metallopeptidase 9\n",
    "* NFKB1: Nuclear Factor Kappa B Subunit 1\n",
    "* HAND1: Heart and Neural Crest Derivatives Expressed 1\n",
    "* CASP3: Caspase 3\n",
    "* PRL: Prolactin\n",
    "\n",
    "\n",
    "* Calcium regulation\n",
    "* CACNA1C: Calcium Voltage-Gated Channel Subunit Alpha1 C\n",
    "* ATP2A2: ATPase Sarcoplasmic/Endoplasmic Reticulum Ca2+ Transporting 2\n",
    "* RYR2: Ryanodine Receptor 2\n",
    "* PLN: - Phospholamban\n",
    "* CALM1: Calmodulin 1\n",
    "* S100A10: S100 Calcium Binding Protein A10\n",
    "* CAMK2D: Calcium/Calmodulin-Dependent Protein Kinase II Delta\n",
    "* ITPR1: Inositol 1,4,5-Trisphosphate Receptor Type 1\n",
    "* CALR: Calreticulin\n",
    "* ATP2B2: ATPase Plasma Membrane Ca2+ Transporting 2\n",
    "* CALM2: Calmodulin 2\n",
    "* CASQ1: Calsequestrin 1\n",
    "* TMEM38B: Transmembrane Protein 38B\n",
    "* ITPKC: Inositol-Tetrakisphosphate 1-Kinase\n",
    "* ATP2B4: ATPase Plasma Membrane Ca2+ Transporting 4\n",
    "* CAMK2B: Calcium/Calmodulin-Dependent Protein Kinase II Beta\n",
    "* SLC8A1: Solute Carrier Family 8 Member A1\n",
    "* PLCD4: ATPase Plasma Membrane Ca2+ Transporting 1\n",
    "* STIM1 Stromal Interaction Molecule 1\n",
    "\n",
    "\n",
    "* NO signaling\n",
    "* NOSi: Nitric Oxide Synthase inducible\n",
    "* NOSe: Nitric Oxide Synthase \n",
    "* NOSn: Nitric Oxide Synthase Neuronal\n",
    "* GUCY1A3: Guanylate Cyclase 1, Soluble, Alpha 3\n",
    "* GUCY1B3: Guanylate Cyclase 1, Soluble, Beta 3\n",
    "* PRKG1: Protein Kinase, cGMP-Dependent, Type I\n",
    "* PRKG2: Protein Kinase, cGMP-Dependent, Type II\n",
    "* CALM1: Calmodulin 1\n",
    "* HMOX1: Heme Oxygenase 1\n",
    "* EDN1: Endothelin 1\n",
    "* CAV1: Caveolin 1\n",
    "* CAV2: Caveolin 2\n",
    "* CAT: Catalase\n",
    "* GPX1: Glutathione Peroxidase 1\n",
    "* SOD1: Superoxide Dismutase 1\n",
    "* SOD2: Superoxide Dismutase 2\n",
    "* GSR: Glutathione-Disulfide Reductase\n",
    "* NRF2: Nuclear Factor, Erythroid 2 Like 2\n",
    "* AKT1: AKT Serine/Threonine Kinase 1\n",
    "* MAPK1: Mitogen-Activated Protein Kinase 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93dd2812",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
